plugins/producer_plugin/include/eosio/producer_plugin/production_pause_vote_tracker.hpp

```diff
@@ -0,0 +1,229 @@
+#pragma once
+
+#include <eosio/chain/thread_utils.hpp>
+#include <fc/time.hpp>
+#include <atomic>
+
+namespace eosio {
+
+class production_pause_vote_tracker {
+public:
+   // By default, start with maximum vote_timeout so that production pause mechanism is effectively disabled.
+   // Pre-conditions: vote_timeout and block_acceptance_tolerance should both not be negative.
+   explicit production_pause_vote_tracker(fc::microseconds block_acceptance_tolerance = fc::milliseconds(250),
+                                          fc::microseconds vote_timeout               = fc::microseconds::maximum())
+      : negative_vote_timeout(), block_acceptance_delta(), latest_other_vote(), latest_producer_vote() {
+      set_vote_timeout(vote_timeout);
+      set_block_acceptance_tolerance(block_acceptance_tolerance);
+   }
+
+   // Can be called concurrently with all member functions except:
+   //    + set_vote_timeout
+   //    + check_pause_status
+   //
+   // Sets the vote timeout duration which drives the production pause status.
+   //
+   // Pre-condition: vote_timeout should not be negative.
+   void set_vote_timeout(fc::microseconds vote_timeout) {
+      if (vote_timeout.count() > 0) {
```

Comment:
I missed this earlier but this should be `>=`.

Reply:
Why? `>=` seems correct to me. Why set it to `-0` ?

Reply:
It wouldn't change the value in state if `vote_timeout` is 0.

We could argue that 0 is not a sensible value to pass in to the function. If we want to go that route instead, then we should update the comment for the pre-condition to say vote_timeout should be positive.